lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -283,58 +283,75 @@ static bool isBound(hw::HWModuleLike op, hw::InstanceGraph &instanceGraph) {
-  // Ensure we have a valid test code path.
-  if (!testBenchDir)
-    return;
-
+// Inline any modules that only have inputs for test code.
+static void inlineInputOnly(hw::HWModuleOp oldMod,
+                            hw::InstanceGraph &instanceGraph,
+                            BindTable &bindTable) {
   // Check if the module only has inputs.
   if (oldMod.getNumOutputs() != 0)
     return;
 
-  // Check if this module has been bound already, and return early if so. This
-  // can happen, for example, if a module is input only, but exists to feed both
-  // asserts and covers. In such cases, the binds end up in the bind file for
-  // the first kind of test code statement we saw, and we expect this to not
-  // cause issues.
-  if (isBound(oldMod, instanceGraph)) {
-    LLVM_DEBUG(oldMod->emitWarning("already bound, skipping for ") << bindFile);
-    return;
-  }
-
-  // Move the module to the test code path.
-  oldMod->setAttr("output_file", testBenchDir);
-
-  // Replace its instances with binds.
+  // Iterate through each of the module's inputs.
   hw::InstanceGraphNode *node = instanceGraph.lookup(oldMod);
-  OpBuilder b = OpBuilder(oldMod.getContext());
+  OpBuilder b(oldMod);
+  bool allInlined = true;
   for (hw::InstanceRecord *use : node->uses()) {
-    OpBuilder::InsertionGuard g(b);
-
+    // If there is no instance, move on.
     hw::HWInstanceLike instLike = use->getInstance();
-    if (!instLike)
+    if (!instLike) {
+      allInlined = false;
       continue;
+    }
 
-    // Mark the instance to not be printed.
+    // If the instance had a symbol, we can't inline it.
     hw::InstanceOp inst = cast<hw::InstanceOp>(instLike.getOperation());
-    hw::HWModuleLike parent = use->getParent()->getModule();
-    inst->setAttr("doNotPrint", b.getBoolAttr(true));
-
-    // Give the instance an inner sym if it needs one.
-    if (!inst.getInnerSym().has_value())
-      inst.setInnerSymAttr(
-          b.getStringAttr("__" + inst.getInstanceName() + "__"));
-
-    // Create the bind.
-    b.setInsertionPointToEnd(
-        &oldMod->getParentOfType<mlir::ModuleOp>()->getRegion(0).front());
-    auto bindOp = b.create<sv::BindOp>(inst.getLoc(), parent.moduleNameAttr(),
-                                       inst.getInnerSymAttr());
-    if (bindFile)
-      bindOp->setAttr("output_file", bindFile);
+    if (inst.getInnerSym().has_value()) {
+      allInlined = false;
+      continue;
+    }
+
+    // Build a mapping from module block arguments to instance inputs.
+    BlockAndValueMapping mapping;
+    assert(inst.getInputs().size() == oldMod.getNumInputs());
+    auto inputPorts = oldMod.getBodyBlock()->getArguments();
+    for (size_t i = 0, e = inputPorts.size(); i < e; ++i)
+      mapping.map(inputPorts[i], inst.getOperand(i));
+
+    // Inline the body at the instantiation site.
+    hw::HWModuleOp instParent =
+        cast<hw::HWModuleOp>(use->getParent()->getModule());
+    OpBuilder::InsertionGuard g(b);
```

Comment:
Not sure if you need to bother resetting the insert location each time

Reply:
I was doing it because I wanted to `clone` the inner ops right into the point where the original instance was instantiated. Maybe not strictly necessary, but seemed like the right place to put them.

Reply:
I think it's a good idea to inline them at the instance op.  What was wondering is if you need to bother with the guard, which will reset the insert position back to, I think, the beginning of the block each iteration.

Reply:
Oh I see what you're saying, since I set the insertion point every time, I guess I can just remove the guard.